### PR TITLE
PXB-3064 Document dictionary cache (8.0)

### PR DESCRIPTION
**Pull request by patrickbirch**
new file:   docs/_static/InnoDB-SDI-explanation.png
	new file:   docs/_static/undo-operations.png
	new file:   docs/dictionary-cache.md
	modified:   docs/how-xtrabackup-works.md

### CODE REVIEW

File: docs/dictionary-cache.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Dictionary cache 
+
+After Percona XtraBackup takes the [backup](./how-xtrabackup-works.md#back-up) phase, in the [prepare](./how-xtrabackup-works.md#prepare) phase, `InnoDB` performs the crash recovery operation, as if the server exited.
```
Comment (alina-derkach-oaza) [inline docs/dictionary-cache.md]:
```suggestion
After Percona XtraBackup takes the [backup](./how-xtrabackup-works.md#back-up), in the [prepare](./how-xtrabackup-works.md#prepare) phase, `InnoDB` performs the crash recovery operation as if the server exited.
```

File: docs/dictionary-cache.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+As a logical operation, Percona XtraBackup applies the undo log on a specific row. When the redo log completes, XtraBackup uses the undo log to roll back changes from uncommitted transactions during the backup.
+
+## Undo log
+
+There are two types of undo log records:
+
+* INSERT
+* UPDATE
+
+An undo log record contains a `table_id`. Percona XtraBackup uses this `table_id` to find the table definition, and that information is used to parse the records on an index page. The transaction rollback reads the undo log records and applies the changes.
```
Comment (alina-derkach-oaza) [inline docs/dictionary-cache.md]: and that information is used to parse the records on an index page. ->  and then uses that information to parse the records on an index page.

File: docs/dictionary-cache.md (hunk 1 above)
Comment (satya-bodapati) [reply docs/dictionary-cache.md]: Let's rewrite this a bit. 

A backup taken Percona XtraBackup must be prepared before being restored. The prepare phase in xtrabackup is equivalent to the InnoDB engine performing a crash recovery.

File: docs/dictionary-cache.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+This method has the following issues:
+
+* Unnecessary IO operations to read SDI pages due to loading the tables that are not required for rollback
+* All tables loaded into the cache increases the time required in the `--prepare` phase
+* Unnecessary tables can lead to out-of-memory errors
+* Huge number of tables and IBD files may cause an exit in the `--prepare` phase
+
+Percona XtraBackup 8.0.33-27 implements a new design and tables are loaded as `evictable`. XtraBackup scans the B-tree data dictionary tables `mysql.indexes` and `mysql.index_partitions` to establish a relationship between the `table_id` and the tablespace(space_id). XtraBackup uses this relationship during transaction rollback. XtraBackup does not load user tables unless there is a transaction rollback on them.
+
+The server evicts tables from the cache when the cache size limit is reached or uses a background master thread.
```
Comment (satya-bodapati) [inline docs/dictionary-cache.md]: A background thread or a PXB main thread handles the cache eviction when the cache size limit is reached.

File: docs/dictionary-cache.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+* INSERT
+* UPDATE
+
+An undo log record contains a `table_id`. Percona XtraBackup uses this `table_id` to find the table definition, and that information is used to parse the records on an index page. The transaction rollback reads the undo log records and applies the changes. 
+
+After initializing the data dictionary engine and the data dictionary cache, the storage engine can request the `table_id`. For example, InnoDB uses the `table_id` (also known as the `se_private_id`) for a table definition. Percona XtraBackup does not behave like a server and does not have access to the data dictionary. XtraBackup initializes the InnoDB engine and uses the `InnoDB table object` (`dict_table_t`) when needed. XtraBackup relies on Serialized Dictionary Information (SDI) which is stored in the tablespace. SDI is a JSON representation of a table.
+
+## Version changes
+
+Prior to Percona XtraBackup 8.0.33-27, XtraBackup loads the SDI from every `.IBD` file and all tables into cache as non-evictable. Making the tables non-evictable essentially disables the LRU cache. Every table remains in memory until the operation ends.
```
Comment (satya-bodapati) [inline docs/dictionary-cache.md]: Prior to 8.0.33-28

File: docs/dictionary-cache.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+Prior to Percona XtraBackup 8.0.33-27, XtraBackup loads the SDI from every `.IBD` file and all tables into cache as non-evictable. Making the tables non-evictable essentially disables the LRU cache. Every table remains in memory until the operation ends. 
+
+This method has the following issues:
+
+* Unnecessary IO operations to read SDI pages due to loading the tables that are not required for rollback
+* All tables loaded into the cache increases the time required in the `--prepare` phase
+* Unnecessary tables can lead to out-of-memory errors
+* Huge number of tables and IBD files may cause an exit in the `--prepare` phase
+
+Percona XtraBackup 8.0.33-27 implements a new design and tables are loaded as `evictable`. XtraBackup scans the B-tree data dictionary tables `mysql.indexes` and `mysql.index_partitions` to establish a relationship between the `table_id` and the tablespace(space_id). XtraBackup uses this relationship during transaction rollback. XtraBackup does not load user tables unless there is a transaction rollback on them.
```
Comment (satya-bodapati) [inline docs/dictionary-cache.md]: 8.0.33-28

File: docs/dictionary-cache.md (hunk 1 above)
Comment (satya-bodapati) [reply docs/dictionary-cache.md]: I also see that we have somethign similar written below (how-xtrabackup works.md)

This is good too. we can use this as is

```
Percona XtraBackup is based on how [`crash recovery`](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_crash_recovery) works. Percona XtraBackup copies the InnoDB data files, which results in data that is internally inconsistent; then performs crash recovery on the files to make a consistent, usable database again.
```

we can say the prepare phase performs the crash recovery and make the database consistent.

File: docs/how-xtrabackup-works.md [hunk 6]
```diff
@@ -1,110 +1,101 @@
...
+
+## Version changes
+
+[Percona XtraBackup 8.0.30-23](release-notes/8.0/8.0.30-23.0.md) adds the `--register-redo-log-consumer` parameter, which is disabled by default. When enabled, this parameter lets Percona XtraBackup register as a redo log consumer at the start of the backup. The server does not remove a redo log that Percona XtraBackup (the consumer) has not yet copied. The consumer reads the redo log and manually advances the log sequence number (LSN). The server blocks the writes during the process. Based on the redo log consumption, the server determines when it can purge the log.  
+
+## Phases of XtraBackup
+
+The phases of an backup and restore process are the following:
+
+* Back up
```
Comment (satya-bodapati) [inline docs/how-xtrabackup-works.md]: Backup (combined)

File: docs/how-xtrabackup-works.md [hunk 7]
```diff
@@ -1,110 +1,101 @@
...
+* Prepare
+* Restore
+
+
+## Back up
```
Comment (satya-bodapati) [inline docs/how-xtrabackup-works.md]: Backup

File: docs/how-xtrabackup-works.md [hunk 8]
```diff
@@ -1,110 +1,101 @@
...
+
+Percona XtraBackup remembers the log sequence number (LSN) when it starts, and then copying away the data files. The backup takes time. If the data files change, the files reflect the database state at different points in time. The server writes the redo log in a circular fashion, and can be reused after a
```
Comment (satya-bodapati) [inline docs/how-xtrabackup-works.md]: From 8.0.30, server no longer writes in a circular fashion. Instead, it creates new files and purges old redo log files.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: docs/dictionary-cache.md (hunk 1 above)
Comment (patrickbirch) [reply docs/dictionary-cache.md]: Rewritten

File: docs/dictionary-cache.md (hunk 4 above)
Comment (patrickbirch) [reply docs/dictionary-cache.md]: updated

File: docs/dictionary-cache.md (hunk 5 above)
Comment (patrickbirch) [reply docs/dictionary-cache.md]: updated

File: docs/dictionary-cache.md (hunk 3 above)
Comment (patrickbirch) [reply docs/dictionary-cache.md]: Rewritten

File: docs/how-xtrabackup-works.md (hunk 6 above)
Comment (patrickbirch) [reply docs/how-xtrabackup-works.md]: combined

File: docs/how-xtrabackup-works.md (hunk 7 above)
Comment (patrickbirch) [reply docs/how-xtrabackup-works.md]: done

File: docs/how-xtrabackup-works.md (hunk 8 above)
Comment (patrickbirch) [reply docs/how-xtrabackup-works.md]: Done

File: docs/dictionary-cache.md (hunk 2 above)
Comment (patrickbirch) [reply docs/dictionary-cache.md]: updated.

File: docs/dictionary-cache.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+## Undo log
+
+There are two types of undo log records:
+
+* INSERT
+* UPDATE
+
+An undo log record contains a `table_id`. Percona XtraBackup uses this `table_id` to find the table definition, and then uses that information to parse the records on an index page. The transaction rollback reads the undo log records and applies the changes. 
+
+After initializing the data dictionary engine and the data dictionary cache, the storage engine can request the `table_id` and uses this ID to fetch the table schema. An index search tuple (key) are created from the table schema and key fields from the undo log record. The server finds the record using the search tuple (key) and performs the undo operation.
```
Comment (satya-bodapati) [inline docs/dictionary-cache.md]: An index search tuple (key) is created..

File: docs/dictionary-cache.md [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+Prior to Percona XtraBackup 8.0.33-28, XtraBackup loads the SDI from every `.IBD` file and all tables into cache as non-evictable. Making the tables non-evictable essentially disables the LRU cache. Every table remains in memory until the operation ends. 
+
+This method has the following issues:
+
+* Unnecessary IO operations to read SDI pages due to loading the tables that are not required for rollback
+* All tables loaded into the cache increases the time required in the `--prepare` phase
+* Unnecessary tables can lead to out-of-memory errors
+* Huge number of tables and IBD files may cause an exit in the `--prepare` phase
+
+Percona XtraBackup 8.0.33-28 implements a new design and tables are loaded as `evictable`. XtraBackup scans the B-tree data dictionary tables `mysql.indexes` and `mysql.index_partitions` to establish a relationship between the `table_id` and the tablespace(space_id). XtraBackup uses this relationship during transaction rollback. XtraBackup does not load user tables unless there is a transaction rollback on them.
```
Comment (satya-bodapati) [inline docs/dictionary-cache.md]: scans the B-tree index of data dictionary*

File: docs/dictionary-cache.md [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+There are two types of undo log records:
+
+* INSERT
+* UPDATE
+
+An undo log record contains a `table_id`. Percona XtraBackup uses this `table_id` to find the table definition, and then uses that information to parse the records on an index page. The transaction rollback reads the undo log records and applies the changes. 
+
+After initializing the data dictionary engine and the data dictionary cache, the storage engine can request the `table_id` and uses this ID to fetch the table schema. An index search tuple (key) are created from the table schema and key fields from the undo log record. The server finds the record using the search tuple (key) and performs the undo operation.
+
+For example, InnoDB uses the `table_id` (also known as the `se_private_id`) for a table definition. Percona XtraBackup does not behave like a server and does not have access to the data dictionary. XtraBackup initializes the InnoDB engine and uses the `InnoDB table object` (`dict_table_t`) when needed. XtraBackup relies on Serialized Dictionary Information (SDI) which is stored in the tablespace. SDI is a JSON representation of a table.
```
Comment (satya-bodapati) [inline docs/dictionary-cache.md]: XtraBackup relies on Serialized Dictionary Information (SDI) **that** is stored in the tablespace. SDI

File: docs/dictionary-cache.md [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+## Version changes
+
+Prior to Percona XtraBackup 8.0.33-28, XtraBackup loads the SDI from every `.IBD` file and all tables into cache as non-evictable. Making the tables non-evictable essentially disables the LRU cache. Every table remains in memory until the operation ends. 
+
+This method has the following issues:
+
+* Unnecessary IO operations to read SDI pages due to loading the tables that are not required for rollback
```
Comment (satya-bodapati) [inline docs/dictionary-cache.md]: Unnecessary IO operations from reading SDI pages to load the tables that are not required for rollback

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: docs/dictionary-cache.md (hunk 9 above)
Comment (patrickbirch) [reply docs/dictionary-cache.md]: fixed.

File: docs/dictionary-cache.md (hunk 11 above)
Comment (patrickbirch) [reply docs/dictionary-cache.md]: updated.

File: docs/dictionary-cache.md (hunk 12 above)
Comment (patrickbirch) [reply docs/dictionary-cache.md]: updated.

Comment (patrickbirch) [reply docs/dictionary-cache.md]: done

File: docs/dictionary-cache.md (hunk 10 above)
Comment (patrickbirch) [reply docs/dictionary-cache.md]: updated

**Review by satya-bodapati — APPROVED** [review]
LGTM!

**Review by alina-derkach-oaza — APPROVED** [review]
LGTM

**Pull request merged**